### ISSUE TITLE
More general unvendoring mechanism

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#3092]
Currently, we unvendor tests via the `build/unvendor-tests` flag. It could be useful to additionally have a more generic unvendoring mechanism to unvendor other sub-components. For instance, some packages have included datasets (sometimes rather large) that it could be useful to unvendor:  geopandas 1MB, scikit-learn 0.5MB, scikit-image: 7.5MB

Maybe it could be a flag of the form,
```yaml
build:
    unvendor:
         - package-name: <regrex or glob>
         - package-name2: <regex or glob 2>
```

Also,
- [ ] ideally, the list of unvendored packages for a given package should be easily discoverable in the  `repodata.json` (including the unvendored tests)
- [ ] we could add another `MetaPathFinder`  similar to [`UnvendoredStdlibFinder`](https://github.com/pyodide/pyodide/blob/b4433e1b4a28f4d904c85bd417f95ea667d329fc/src/py/_pyodide/_importhook.py#L145) which for a given ImporError, looks at the path, checks it's in a known `top_level` fields, and if it raises a meaningful error about how to install the missing component with micropip. This would improve a lot the usability of unvendored packages.  
- [ ] The previous step would require that the `top_level` entries are correctly auto-determined from the unvendored packages (which is likely not the case currently). Related https://github.com/pyodide/pyodide/pull/3006

**Comment by ryanking13** [comment]
It sounds good to me. 

> * looks at the path, checks it's in a known `top_level` fields

I think unvendored components are mostly not top_level directories, so I don't think it is a good idea to use top_level key for that usage. Perhaps we need a separate key for that.

**Comment by rth** [comment]
>  unvendored components are mostly not top_level directories, so I think we need a separate key for that usage.

Good point, that they are not top level. In the end though we don't really care much whether these paths are top level or not,  I think? All we care is that we can use it to have the import name -> package name mapping. And it would be better to have 1 field for this rather than two. So maybe having an `package>imports` field would be actually more general than `package>top_level`. Particularly since it's not part of a stable release yet.  WDYT?

**Comment by ryanking13** [comment]
> > unvendored components are mostly not top_level directories, so I think we need a separate key for that usage.
> 
> Good point, that they are not top level. In the end though we don't really care much whether these paths are top level or not, I think? All we care is that we can use it to have the import name -> package name mapping. And it would be better to have 1 field for this rather than two. So maybe having an `package>imports` field would be actually more general than `package>top_level`. Particularly since it's not part of a stable release yet. WDYT?

Yes, that is actually a good point. In that case, we may need find a way to distinguish vendored and unvendored parts of the packages to tell correct error messages. For example. `import regex` should tell that `loadPackage("regex")` is required, while `import regex.test` should tell that `loadPackage("regex-test")` is required.
